File: app/src/main/java/com/balch/auctionbrowser/auction/AuctionPresenter.java
/*
 * Author: Balch
 * Created: 9/4/14 12:26 AM
 *
 * This file is part of MockTrade.
 *
 * MockTrade is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * MockTrade is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MockTrade.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2014
 */

package com.balch.auctionbrowser.auction;

import android.app.LoaderManager;
import android.content.AsyncTaskLoader;
import android.content.Context;
import android.content.Loader;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.balch.android.app.framework.BaseAppCompatActivity;
import com.balch.android.app.framework.BasePresenter;
import com.balch.auctionbrowser.AuctionApplication;
import com.balch.auctionbrowser.R;
import com.balch.auctionbrowser.note.Note;
import com.balch.auctionbrowser.note.NoteEditDialog;
import com.balch.auctionbrowser.note.NotesModel;

import java.util.List;
import java.util.Map;

public class AuctionPresenter extends BasePresenter<AuctionApplication>
        implements LoaderManager.LoaderCallbacks<AuctionPresenter.AuctionData> {
    private static final String TAG = AuctionPresenter.class.getSimpleName();

    protected final AuctionView auctionView;
    protected final EBayModel auctionModel;
    protected final NotesModel notesModel;

    protected static final int AUCTION_LOADER_ID = 0;
    protected static final int AUCTION_FETCH_COUNT = 30;

    protected boolean isLoadFinished = false;

    // keep in sync with auction_sort_col string array
    private final String[] sortColumns =
            new String[]{"BestMatch", "EndTimeSoonest", "PricePlusShippingLowest"};

    // TODO: Serialize this so we can recover on Activity reload
    protected int currentPage = 1;
    protected int sortPosition = 0;
    protected long totalPages = -1;


    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    public AuctionPresenter(AuctionView auctionView, EBayModel auctionModel, NotesModel notesModel) {
        this.auctionView = auctionView;
        this.auctionModel = auctionModel;
        this.notesModel = notesModel;
    }

    @Override
    public void initialize(Bundle savedInstanceState) {
        this.auctionView.setMainViewListener(new AuctionView.MainViewListener() {
            @Override
            public void onLoadMore(int currentPage) {
                boolean hasMore = ((totalPages == -1) ||
                                    (currentPage < totalPages));
                if (isLoadFinished && hasMore) {
                    auctionView.showBusy();
                    AuctionPresenter.this.currentPage = currentPage;
                    loaderManager.restartLoader(AUCTION_LOADER_ID, null, AuctionPresenter.this).forceLoad();
                }
            }

            @Override
            public void onChangeSort(int position) {
                if (isLoadFinished) {
                    sortPosition = position;
                    currentPage = 1;
                    totalPages = -1;

                    auctionView.clearAuctions();
                    loaderManager.restartLoader(AUCTION_LOADER_ID, null, AuctionPresenter.this).forceLoad();
                }
            }

            @Override
            public void onClickNoteButton(final Auction auction) {
                NoteEditDialog dialog = new NoteEditDialog();
                Bundle args = new Bundle();

                final Note note = auctionView.getNote(auction);
                if (note != null) {
                    args.putString(NoteEditDialog.ARG_NOTE, note.getNote());
                    args.putString(NoteEditDialog.ARG_TITLE, auction.getTitle());
                    dialog.setArguments(args);
                    dialog.setNoteEditDialogListener(new NoteEditDialog.NoteEditDialogListener() {
                        @Override
                        public void onSave(String text) {
                            note.setNote(text);
                            notesModel.update(note);
                        }

                        @Override
                        public void onClear() {
                            notesModel.delete(note);
                            auctionView.clearNote(auction);
                        }
                    });
                    dialog.show(((BaseAppCompatActivity) auctionView.getContext()).getSupportFragmentManager(),
                            "NoteEditDialog");
                }
            }

            @Override
            public void onClickAuction(final Auction auction) {
                AuctionDetailDialog dialog = new AuctionDetailDialog();
                Bundle args = new Bundle();

                final Note note = auctionView.getNote(auction);
                if (note != null) {
                    args.putString(AuctionDetailDialog.ARG_NOTE, note.getNote());
                }
                args.putSerializable(AuctionDetailDialog.ARG_AUCTION, auction);
                dialog.setArguments(args);
                dialog.setNoteDetailDialogListener(new AuctionDetailDialog.NoteDetailDialogListener() {
                    @Override
                    public void onSave(String text) {
                        if (note == null) {
                            Note note1 = new Note();
                            note1.setNote(text);
                            note1.setItemId(auction.getItemId());
                            notesModel.insert(note1);
                            auctionView.addNote(auction, note1);
                        } else {
                            note.setNote(text);
                            notesModel.update(note);
                        }
                    }

                    @Override
                    public void onClear() {
                        if (note != null) {
                            notesModel.delete(note);
                            auctionView.clearNote(auction);
                        }
                    }
                });
                dialog.show(((BaseAppCompatActivity) auctionView.getContext()).getSupportFragmentManager(),
                        "NoteDetailDialog");
            }

        });

        this.auctionView.setSortStrings(R.array.auction_sort_col);
        this.auctionView.showBusy();
        this.loaderManager.initLoader(AUCTION_LOADER_ID, null, this).forceLoad();
    }

    @Override
    public Loader<AuctionData> onCreateLoader(int id, Bundle args) {
        return new MemberLoader(this.application, this.currentPage, this.sortColumns[this.sortPosition],
                this.auctionModel, this.notesModel);
    }

    @Override
    public void onLoadFinished(Loader<AuctionData> loader, AuctionData data) {
        auctionView.hideBusy();

        if (data.auctions != null) {
            if (totalPages == -1) {
                totalPages = data.totalPages;
            }
            auctionView.addAuctions(data.auctions, data.notes);
        } else {
            Toast.makeText(getApplication(), R.string.error_auction_get, Toast.LENGTH_LONG).show();
        }
        
        isLoadFinished = true;
    }

    @Override
    public void onLoaderReset(Loader<AuctionData> loader) {

    }

    protected static class AuctionData {
        protected List<Auction> auctions;
        protected Map<Long, Note> notes;
        protected int totalPages;
    }

    protected static class MemberLoader extends AsyncTaskLoader<AuctionData> {
        protected int currentPage;
        protected EBayModel auctionModel;
        protected NotesModel notesModel;
        protected String sortOrder;

        public MemberLoader(Context context, int currentPage, String sortOrder,
                            EBayModel auctionModel, NotesModel notesModel) {
            super(context);
            this.currentPage = currentPage;
            this.sortOrder = sortOrder;
            this.auctionModel = auctionModel;
            this.notesModel = notesModel;
        }

        @Override
        public AuctionData loadInBackground() {
            AuctionData auctionData = new AuctionData();
            try {

                EBayModel.AuctionInfo info = this.auctionModel.getAuctions(
                        "multi-rotor",
                        currentPage  * AUCTION_FETCH_COUNT,
                        AUCTION_FETCH_COUNT,
                        sortOrder);

                auctionData.totalPages = info.totalPages;
                auctionData.auctions = info.auctions;
                if (auctionData.auctions != null) {
                    auctionData.notes = this.notesModel.getNotes(auctionData.auctions);
                }
            } catch (Exception ex) {
                Log.e(TAG, "error getting auctions", ex);
            }

            return auctionData;
        }
    }

}

File: app/src/main/java/com/balch/auctionbrowser/auction/AuctionPresenter.kt
/*
 * Author: Balch
 *
 * This file is part of AuctionBrowser.
 *
 * AuctionBrowser is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AuctionBrowser is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MockTrade.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2018
 *
 */

package com.balch.auctionbrowser.auction

import android.annotation.SuppressLint
import android.os.Bundle
import android.widget.SearchView
import androidx.annotation.VisibleForTesting
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import androidx.paging.PagedList
import com.balch.auctionbrowser.auction.model.Auction
import com.balch.auctionbrowser.auction.model.EBayRepository
import com.balch.auctionbrowser.base.BasePresenter
import com.balch.auctionbrowser.base.NetworkState
import com.balch.auctionbrowser.dagger.ActivityScope
import com.balch.auctionbrowser.note.Note
import com.balch.auctionbrowser.note.NotesRepository
import com.uber.autodispose.lifecycle.LifecycleScopeProvider
import com.uber.autodispose.lifecycle.autoDisposable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

@ActivityScope
class AuctionPresenter
@Inject constructor(override val view: AuctionView,
                    private val auctionViewModel: AuctionViewModel,
                    private val notesRepository: NotesRepository,
                    private val fragmentManager: FragmentManager,
                    private val lifecycleOwner: LifecycleOwner,
                    private val auctionAdapter: AuctionAdapter,
                    private val scope: LifecycleScopeProvider<*>
                    ) : BasePresenter() {

    var searchView: SearchView? = null
        set(value) {
            field = value
            searchView!!.setQuery(auctionViewModel.searchText, false)
        }

    @SuppressLint("VisibleForTests")
    override fun initialize(savedInstanceState: Bundle?) {

        view.setAuctionAdapter(auctionAdapter)

        auctionViewModel.auctionData.observe(lifecycleOwner, Observer<PagedList<Auction>> {
            auctionAdapter.submitList(it)
        })

        auctionViewModel.networkState.observe(lifecycleOwner, Observer<NetworkState> {
            auctionAdapter.setNetworkState(it)
        })

        auctionAdapter.onClickAuction
                .autoDisposable(scope)
                .subscribe({ auction -> showDetail(auction) },
                        { throwable -> Timber.e(throwable, "onClickAuction error") })

        auctionAdapter.onClickNote
                .autoDisposable(scope)
                .subscribe({ auction -> showDetail(auction) },
                        { throwable -> Timber.e(throwable, "onClickNote error") })
    }

    fun doSearch(keyword: String) {
        searchView?.clearFocus()

        auctionViewModel.loadAuctions(keyword, auctionViewModel.sortColumn)
    }

    internal fun sortAuctions(sortColumn: EBayRepository.SortColumn) {
        auctionViewModel.loadAuctions(auctionViewModel.searchText, sortColumn)
    }

    @SuppressLint("VisibleForTests")
    private fun showDetail(auction: Auction) {
        val note = auction.note

        val dialog = AuctionDetailDialog.newInstance(auction, note)

        dialog.onClearNote
                .autoDisposable(scope)
                .subscribe({ clearNote(auction, note) },
                        { throwable -> Timber.e(throwable, "clearNote error") })

        dialog.onSaveNote
                .autoDisposable(scope)
                .subscribe({ text -> saveNote(auction, note, text) },
                        { throwable -> Timber.e(throwable, "saveNote error") })

        dialog.show(fragmentManager, "AuctionDetailDialog")
    }

    @VisibleForTesting
    fun saveNote(auction: Auction, note: Note?, text: String) {
        if (note == null) {
            val newNote = Note(auction.itemId, text)
            notesRepository.insert(newNote)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .autoDisposable(scope)
                .subscribe({
                    auction.note = newNote
                    auctionAdapter.notifyDataSetChanged()},
                        { throwable -> Timber.e(throwable, "insertNote error") })
        } else {
            note.noteText = text
            notesRepository.update(note)
                .subscribeOn(Schedulers.io())
                .autoDisposable(scope)
                .subscribe({ /* no-op */ },
                        { throwable -> Timber.e(throwable, "updateNote error") })
        }
    }


    fun clearNote(auction: Auction, note: Note?) {
        if (note != null) {
            notesRepository.delete(note)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .autoDisposable(scope)
                    .subscribe({ auction.note = null
                                 auctionAdapter.notifyDataSetChanged()},
                            { throwable -> Timber.e(throwable, "deleteNote error") })
        }
    }

    override fun cleanup() {
        view.cleanup()
    }
}